crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -560,12 +569,23 @@ impl AuthorityStorePruner {
     }
 
     fn smoothed_max_eligible_checkpoint_number(
+        checkpoint_store: &Arc<CheckpointStore>,
+        mut max_eligible_checkpoint: CheckpointSequenceNumber,
         pruned_checkpoint: CheckpointSequenceNumber,
-        max_eligible_checkpoint: CheckpointSequenceNumber,
+        epoch_id: EpochId,
         epoch_duration_ms: u64,
-    ) -> CheckpointSequenceNumber {
+        num_epochs_to_retain: u64,
+    ) -> anyhow::Result<CheckpointSequenceNumber> {
+        if epoch_id < num_epochs_to_retain {
+            return Ok(0);
+        }
+        let last_checkpoint_in_epoch = checkpoint_store
+            .get_epoch_last_checkpoint(epoch_id - num_epochs_to_retain)?
```

Comment:
@mystenmark: just a quick sanity check. This watermark should be reliable, right?